src/relay/charpane.ash

```diff
@@ -4532,112 +4532,112 @@ buffer addGroup(string[int] rowHTMLs, string className) {
+				styleInfo[pLevel,i[pLevel]] = s;
+				read_style = false;
+			} else {
+				switch(s) {
+					case "": break;
+					case ":":
+					//next token tells us how wide this last element should be
+						read_style = true;
+						break;
+					case "|":
+					//next element will be horizontally adjacent to this one
+						i[pLevel]++;
+						break;
+					case ",":
+					//next element will start a new row below this one
+						rowHTML[pLevel,j[pLevel]] = addRow(bricks[pLevel], styleInfo[pLevel]);
+						i[pLevel]=0;
+						j[pLevel]++;
+						bricks[pLevel]={};
+						styleInfo[pLevel]={};
+						break;
+					case "{":
+					case "(":
+					//parenthesis depth increased. Pause current group parsing and start a new one.
+						pLevel++;
+						i[pLevel] = 0;
+						j[pLevel] = 0;
+						bricks[pLevel] = {};
+						styleInfo[pLevel] = {};
+						break;
+					case ")":
+					//vertical group complete. Add it to the list of bricks in the previous parenthesis depth and continue there.
+						rowHTML[pLevel,j[pLevel]] = addRow(bricks[pLevel], styleInfo[pLevel]);
+						bricks[pLevel-1,i[pLevel-1]] = addGroup(rowHTML[pLevel],"brick_column");
+						i[pLevel] = 0;
+						j[pLevel] = 0;
+						bricks[pLevel] = {};
+						styleInfo[pLevel] = {};
+						pLevel--;
+						break;
+					case "}":
+					//horizontal row complete. Add it to the list of bricks in the previous parenthesis depth and continue there.
+						rowHTML[pLevel,j[pLevel]] = addRow(bricks[pLevel], styleInfo[pLevel]);
+						bricks[pLevel-1,i[pLevel-1]] = addGroup(rowHTML[pLevel],"brick_row");
+						i[pLevel] = 0;
+						j[pLevel] = 0;
+						bricks[pLevel] = {};
+						styleInfo[pLevel] = {};
+						pLevel--;
+						break;
+					case "toolbar":
+					case "header":
+					case "footer":
+						break;	//Special Bricks that are inserted manually in the correct places
+					default:
+						if(chitBricks contains s && chitBricks[s] != "") {
+							bricks[pLevel,i[pLevel]] = chitBricks[s];
+						}
+						break;
+				}
 			}
 		}
+		if(pLevel != 0)
+			append(result, "<span>Malformed chit layout (mismatched parentheses): " + layout + "<span>");
+		while(pLevel > 0){
+			rowHTML[pLevel,j[pLevel]] = addRow(bricks[pLevel], styleInfo[pLevel]);
+			bricks[pLevel-1,i[pLevel-1]] = addGroup(rowHTML[pLevel],"brick_row");
+			i[pLevel] = 0;
+			j[pLevel] = 0;
+			bricks[pLevel] = {};
+			styleInfo[pLevel] = {};
+			pLevel--;
+		}
+		rowHTML[pLevel,j[pLevel]] = addRow(bricks[pLevel], styleInfo[pLevel]);
+		append(result, addGroup(rowHTML[pLevel],"brick_row"));
+		return result;
 	}
-	rowHTML[pLevel,j[pLevel]] = addRow(bricks[pLevel], styleInfo[pLevel]);
-	append(result, addGroup(rowHTML[pLevel],"brick_row"));
-	return result;
-}
-
-buffer addBricks(string layout) {
+	
 //receive chit.__.layout string, tokenize it, and send it for processing.
 	string[int] tokens = tokenize(layout, $strings[\,,|,(,),:,{,}]);
-	return addBricks(tokens);
+	return addBricksHelper(tokens);
```

Comment:
Considering this helper is only used in one place as far as I can see, it doesn't really seem needed... But again, that's not a big deal at all.

Reply:
It was more work to refactor it. If I wrote it again, I would probably not write it this way. I pulled in into the main addBricks function so that it has access to the "layout" parameter (used in error message) without having to pass it along as another parameter. No strong feelings about whether it needs to change, so I'm inclined to leave it.

Reply:
Fair enough!